pallets/deitos/src/lib.rs

```diff
@@ -113,54 +135,180 @@ pub mod pallet {
     pub enum Error<T> {
         /// The value retrieved was `None` as no value was previously set.
-        NoneValue,
-        StorageOverflow,
         NonExistentStorageValue,
+        /// Math overflow
+        Overflow,
+        /// Insufficient storage
+        InsufficientStorage,
+        /// On going agreements
+        OnGoingAgreements,
     }
 
     #[pallet::call]
     impl<T: Config> Pallet<T> {
         #[pallet::call_index(0)]
         #[pallet::weight(T::WeightInfo::register_ip())]
         /* to add more parameters*/
-        pub fn register_ip(origin: OriginFor<T>, price_unit: BalanceOf<T>) -> DispatchResult {
+        pub fn register_ip(
+            origin: OriginFor<T>,
+            price_storage_per_block: BalanceOf<T>,
+            total_storage: Storage,
+        ) -> DispatchResult {
             // Check that the extrinsic was signed and get the signer.
-            let who = ensure_signed(origin)?;
+            let ip = ensure_signed(origin)?;
+            let bounding_amount = IPBoundingAmount::<T>::get()?;
+
+            T::Currency::hold(&HoldReason::Transfer.into(), &ip, bounding_amount)?;
+
+            let ip_details = InfraProviderDetails::<T> {
+                price_storage_per_block,
+                total_storage,
+                reserved_storage: Zero::zero(),
+                status: IPStatus::Validating,
+                active_agreements: BoundedVec::new(),
+            };
+
+            InfrastructureProvider::<T>::insert(&ip, ip_details);
+
+            Self::deposit_event(Event::IPRegistered {
+                ip,
+                price_storage_per_block,
+                total_storage,
+            });
 
             Ok(())
         }
 
+        /// This is a temporary call to manage the IP status.
+        /// Statuses updates should be done automatically after an environment software check.
         #[pallet::call_index(1)]
-        #[pallet::weight(T::WeightInfo::update_ip_details())]
-        pub fn update_ip_details(origin: OriginFor<T>, price_unit: BalanceOf<T>) -> DispatchResult {
-            // Check that the extrinsic was signed and get the signer.
-            let who = ensure_signed(origin)?;
+        #[pallet::weight(T::WeightInfo::register_ip())]
+        pub fn update_ip_status(
+            origin: OriginFor<T>,
+            provider: AccountIdLookupOf<T>,
+            status: IPStatus,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+            let bounding_amount = IPBoundingAmount::<T>::get()?;
+            let ip = T::Lookup::lookup(provider)?;
+
+            InfrastructureProvider::<T>::try_mutate(
+                &ip,
+                |ip_details| -> Result<_, DispatchError> {
+                    let ip_details = ip_details
+                        .as_mut()
+                        .ok_or(Error::<T>::NonExistentStorageValue)?;
+
+                    ip_details.status = status.clone();
+
+                    Self::deposit_event(Event::IPStatusChanged {
```

Comment:
Is it possible to move event triggering out of closure?

Reply:
hmmm i can change it of course but i would be interested in to know what would be the reason.

Reply:
the event triggering isn't a part of storage item update logic. I also prefer to keep the nested code minimal because of readability.